time test failure

This failure has been occurring intermittently in CI (e.g. https://app.circleci.com/pipelines/github/nanovms/nanos/727/workflows/fb206d54-7f70-46d4-86e2-d5dc9d235b53/jobs/5043)

```
test_posix_timers: signal test passed
test_itimers: testing itimer interface
test_itimers: starting itimer test
test_itimers: starting: which 0, interval 1000 nsec, one-shot
test_itimers: test passed; delta 245111 nsec
test_itimers: starting: which 0, interval 1000 nsec, periodic
test_itimers: test passed; delta 49470 nsec
test_itimers: starting: which 0, interval 1000000 nsec, one-shot
test_itimers: test passed; delta 197433 nsec
test_itimers: starting: which 0, interval 1000000 nsec, periodic
test_itimers: test passed; delta 83756 nsec
test_itimers: starting: which 0, interval 1000000000 nsec, one-shot
test_itimers: test passed; delta 3553654 nsec
test_itimers: starting: which 0, interval 1000000000 nsec, periodic
test_itimers: test passed; delta 2854846 nsec
test_itimers: starting: which 0, interval 2000000000 nsec, one-shot
test_itimers: test passed; delta 5243714 nsec
test_itimers: starting: which 0, interval 2000000000 nsec, periodic
test_itimers: test passed; delta 23520440 nsec
test_itimers: itimer test passed
test_cputime: test_cputime: proc_delta 281541: thread_delta 630392
make[3]: *** [Makefile:375: run-noaccel] Error 1
make[3]: Leaving directory '/home/circleci/project/platform/pc'
make[2]: *** [Makefile:92: run-noaccel] Error 2
make[2]: Leaving directory '/home/circleci/project'
make[1]: *** [Makefile:83: runtime-tests-noaccel] Error 2
make[1]: Leaving directory '/home/circleci/project'
make: *** [Makefile:76: test-noaccel] Error 2
```

The check "proc_delta < thread_delta" fails because the accounted process time is less than the accounted thread time. This may be related to recent changes involving thread pause/resume (at which points some per-thread time bookkeeping occurs).

--- a/src/aarch64/machine.h
+++ b/src/aarch64/machine.h
@@ -160,11 +160,6 @@ static inline word fetch_and_add(word *target, word num)
     return __sync_fetch_and_add(target, num);
 }
 
-static inline u64 fetch_and_add_64(u64 *target, u64 num)
-{
-    return __sync_fetch_and_add(target, num);
-}
-
 static inline void kern_pause(void)
 {
   // XXX

--- a/src/unix/thread.c
+++ b/src/unix/thread.c
@@ -151,9 +151,8 @@ static inline void run_thread_frame(thread t)
     check_stop_conditions(t);
     kern_lock(); // xx - make thread entry a separate exclusion region for performance
     thread old = current;
-    set_current_thread((nanos_thread)t);
-    ftrace_thread_switch(old, current);    /* ftrace needs to know about the switch event */
     thread_enter_user(t);
+    ftrace_thread_switch(old, t);    /* ftrace needs to know about the switch event */
 
     /* cover wake-before-sleep situations (e.g. sched yield, fs ops that don't go to disk, etc.) */
     t->blocked_on = 0;

--- a/src/unix/unix.c
+++ b/src/unix/unix.c
@@ -331,7 +331,6 @@ process create_process(unix_heaps uh, tuple root, filesystem fs)
     create_stdfiles(uh, p);
     init_threads(p);
     p->syscalls = linux_syscalls;
-    p->utime = p->stime = 0;
     init_sigstate(&p->signals);
     zero(p->sigactions, sizeof(p->sigactions));
     p->posix_timer_ids = create_id_heap(h, h, 0, U32_MAX, 1);
@@ -353,7 +352,6 @@ void thread_enter_system(thread t)
     if (!t->sysctx) {
         timestamp here = now(CLOCK_ID_MONOTONIC);
         timestamp diff = here - t->start_time;
-        fetch_and_add_64(&t->p->utime, diff);
         t->utime += diff;
         t->start_time = here;
         t->sysctx = true;
@@ -365,14 +363,11 @@ void thread_pause(thread t)
 {
     if (get_current_thread() != &t->thrd)
         return;
-    process p = t->p;
     timestamp diff = now(CLOCK_ID_MONOTONIC) - t->start_time;
     if (t->sysctx) {
-        fetch_and_add_64(&p->stime, diff);
         t->stime += diff;
     }
     else {
-        fetch_and_add_64(&p->utime, diff);
         t->utime += diff;
     }
     set_current_thread(0);
@@ -404,7 +399,7 @@ static timestamp stime_updated(thread t)
 
 timestamp proc_utime(process p)
 {
-    timestamp utime = p->utime;
+    timestamp utime = 0;
     thread t;
     vector_foreach(p->threads, t)
         if (t)
@@ -414,7 +409,7 @@ timestamp proc_utime(process p)
 
 timestamp proc_stime(process p)
 {
-    timestamp stime = p->stime;
+    timestamp stime = 0;
     thread t;
     vector_foreach(p->threads, t)
         if (t)

--- a/src/unix/unix_internal.h
+++ b/src/unix/unix_internal.h
@@ -413,7 +413,6 @@ typedef struct process {
     rangemap          vmaps;    /* process mappings */
     vmap              stack_map;
     vmap              heap_map;
-    timestamp         utime, stime;
     struct sigstate   signals;
     struct sigaction  sigactions[NSIG];
     id_heap           posix_timer_ids;

--- a/src/x86_64/machine.h
+++ b/src/x86_64/machine.h
@@ -50,8 +50,7 @@ static inline void memory_barrier(void)
 
 static inline word fetch_and_add(word *variable, word value)
 {
-    asm volatile("lock; xadd %0, %1" : "+r" (value), "+m" (*variable) :: "memory", "cc");
-    return value;
+    return __sync_fetch_and_add(variable, value);
 }
 
 static inline void atomic_set_bit(u64 *target, u64 bit)
@@ -62,11 +61,6 @@ static inline void atomic_set_bit(u64 *target, u64 bit)
 static inline void atomic_clear_bit(u64 *target, u64 bit)
 {
     asm volatile("lock btrq %1, %0": "+m"(*target):"r"(bit) : "memory");
-}
-
-static inline u64 fetch_and_add_64(u64 *target, u64 num)
-{
-    return __sync_fetch_and_add(target, num);
 }
 
 static inline void kern_pause(void)

--- a/test/runtime/time.c
+++ b/test/runtime/time.c
@@ -210,14 +210,15 @@ static void test_cputime(void)
         thread_delta = delta_nsec(&thread_ts, &tmp_ts);
         if (thread_delta < 0)
             fail_error("%s: thread_delta %lld\n", __func__, thread_delta);
+        thread_ts = tmp_ts;
         if (clock_gettime(CLOCK_PROCESS_CPUTIME_ID, &tmp_ts) < 0)
             fail_perror("clock_gettime(CLOCK_PROCESS_CPUTIME_ID)");
         proc_delta = delta_nsec(&proc_ts, &tmp_ts);
         if (proc_delta < 0)
             fail_error("%s: proc_delta %lld\n", __func__, proc_delta);
-        if (proc_delta < thread_delta)
-            fail_error("%s: proc_delta %lld: thread_delta %lld\n", __func__,
-                       proc_delta, thread_delta);
+        proc_ts = tmp_ts;
+        if (delta_nsec(&thread_ts, &proc_ts) < 0)
+            fail_error("%s: process CPU time < thread CPU time\n", __func__);
     } while ((thread_delta == 0) || (proc_delta == 0));
 }
 

